astropy/modeling/mappings.py

```diff
@@ -182,3 +183,115 @@ def inverse(self):
+        dimensionless input).  The second element is the unit that the model will
+        return.
+    input_units_equivalencies : dict, optional
+        Default equivalencies to apply to input values.  If set, this should be a
+        dictionary where each key is a string that corresponds to one of the
+        model inputs.
+    name : str, optional
+        A human-friendly name associated with this model instance
+        (particularly useful for identifying the individual components of a
+        compound model).
+    meta : dict-like, optional
+        Free-form metadata to associate with this model.
+
+    Examples
+    --------
+
+    >>> from astropy.modeling.models import Polynomial1D, UnitsMapping
+    >>> from astropy import units
+    >>> poly = Polynomial1D(1, c0=1, c1=2)
+    >>> model = UnitsMapping(((units.m, None),)) | poly
+    >>> model = model | UnitsMapping(((None, units.s),))
+    >>> model(units.Quantity(10, units.m))  # doctest: +FLOAT_CMP
+    <Quantity 21. s>
+    """
+    def __init__(self, mapping, input_units_equivalencies=None, name=None, meta=None):
+        self._mapping = mapping
+
+        # These attributes are read and handled by Model
+        self._input_units_strict = True
+        self.input_units_equivalencies = input_units_equivalencies
+
+        super().__init__(name=name, meta=meta)
+
+        # Can't invoke this until after super().__init__, since
+        # we need self.inputs and self.outputs to be populated.
+        self._rebuild_units()
+
+    def _rebuild_units(self):
+        self._input_units = {input_name: input_unit for input_name, (input_unit, _) in zip(self.inputs, self.mapping)}
+        self._return_units = {output_name: return_unit for output_name, (_, return_unit) in zip(self.outputs, self.mapping)}
+
+    @property
+    def n_inputs(self):
+        return len(self._mapping)
+
+    @property
+    def n_outputs(self):
+        return len(self._mapping)
+
+    @property
+    def inputs(self):
+        return super().inputs
+
+    @inputs.setter
+    def inputs(self, value):
+        super(UnitsMapping, self.__class__).inputs.fset(self, value)
+        self._rebuild_units()
+
+    @property
+    def outputs(self):
+        return super().outputs
+
+    @outputs.setter
+    def outputs(self, value):
+        super(UnitsMapping, self.__class__).outputs.fset(self, value)
+        self._rebuild_units()
+
+    @property
+    def input_units(self):
+        return self._input_units
+
+    @property
+    def return_units(self):
+        return self._return_units
+
+    @property
+    def mapping(self):
+        return self._mapping
+
+    def evaluate(self, *args):
+        result = []
+        for arg, (_, return_unit) in zip(args, self.mapping):
+            if isinstance(arg, Quantity):
+                value = arg.value
+            else:
+                value = arg
+            result.append(Quantity(value, return_unit, subok=True))
```

Comment:
Has this been tested with a unit that is compatible with the defined unit but not the same (e.g., specifying the input unit to be meters, but providing an input argument in cm)? I'm not sure I see how the implicit conversion is done here since I would expect use of the .to() method to get the value in the specified units (more specifically, if arg = 1000 * u.cm then arg.to(u.m).value should give the value in meters), but I may be misunderstanding something going on implicitly.

Reply:
The machinery of Model takes care of that for us -- since I set `_input_units_strict = True`, it will convert to meters before passing the input to evaluate.  I'll add a line to the example to demonstrate.